p8e-util/src/main/kotlin/io/p8e/crypto/Pen.kt

```diff
@@ -64,19 +64,61 @@ class Pen(
             .orThrow { IllegalStateException("can't verify signature - public cert may not match private key.") }
     }
 
-    override fun sign(): ByteArray = signature.sign()
-
-    override fun update(data: ByteArray) { signature.update(data) }
+    override fun sign(): ByteArray {
+        signature.update(aggregatedData)
+        return signature.sign()
+    }
 
-    override fun update(data: ByteArray, off: Int, len: Int) { signature.update(data, off, len) }
+    override fun update(data: ByteArray) = signature.update(data)
+
+    override fun update(data: ByteArray, off: Int, res: Int) {
+        // If off is less then res, these are the data that we care about.
+        if(off < res) {
+            if(!verifying) {
+                val dataSample = data.copyOfRange(off, off+res)
+                aggregatedData = dataSample
+            } else {
+                /**
+                 * The downstream (data verification) chunks the data into data size of 8192.
+                 * The data needs to be aggregated to its signing size of 32768 before the
+                 * data can be validated.
+                 */
+                if(objSizeIndexer == OBJECT_SIZE_BYTES) {
+                    objSizeIndexer = (off + res)
+                    aggregatedData = null
+                    aggregatedData = if (aggregatedData == null) {
+                        data.copyOfRange(off, off + res)
+                    } else {
+                        aggregatedData?.plus(data.copyOfRange(off, off + res))
+                    }
+                } else {
+                    objSizeIndexer += (off + res)
+                    aggregatedData = aggregatedData?.plus(data.copyOfRange(off, off + res))
+                }
```

Comment:
This is currently used to sign and verify data via the SignatureInputStream for SmartKey, moved this implementation into the Pen class to keep the way of signing/verification consistent between all signer types.

Reply:
this class seems like it's very stateful. Does every sign/verify get its own `Pen` class? If it were to be shared across threads it seems like it could have problems signing/verifying two things concurrently.

Reply:
Every affiliate executing something would have to grab their own signer class (Pen in this case), so we might be ok, unless im missing a bigger picture.